### PR TITLE
Add TBB-based counter for global read-write lock

### PR DESCRIPTION
**Pull request by bendavid**
Adds TBB-based counter types for global read-write locks and use one by default when TBB is available.

This gives a 7% speedup in a medium sized test filling histograms from CMS NanoAOD with RDF and 256 threads.

This implements the last part of the improvements originally discussed in https://github.com/root-project/root/pull/6919

### CODE REVIEW

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
@bendavid ```
2:09:55    In file included from input_line_7:20:
12:09:55    In file included from C:/build/workspace/root-pullrequests-build/build/include\ROOT/TReentrantRWLock.hxx:26:
12:09:55    In file included from C:/build/workspace/root-pullrequests-build/build/include\tbb/enumerable_thread_specific.h:23:
12:09:55    In file included from C:/build/workspace/root-pullrequests-build/build/include\tbb/atomic.h:42:
12:09:55    In file included from C:/build/workspace/root-pullrequests-build/build/include\tbb/tbb_machine.h:113:
12:09:55    In file included from C:/build/workspace/root-pullrequests-build/build/include\tbb/tbb_stddef.h:119:
12:09:55  C:/build/workspace/root-pullrequests-build/build/include\tbb/internal/_tbb_windef.h(28,2): error GAAAB8493: TBB requires linkage with multithreaded C/C++ runtime library.        Choose multithreaded DLL runtime in project settings, or use /MD[d] compiler switch. [C:\build\workspace\root-pullrequests-build\build\core\thread\G__Thread.vcxproj]
12:09:55    #error TBB requires linkage with multithreaded C/C++ runtime library. \
12:09:55     ^
12:09:56  CUSTOMBUILD : error : C:/build/workspace/root-pullrequests-build/build/bin/rootcling_stage1.exe: compilation failure (C:/build/workspace/root-pullrequests-build/build/bin/libThreada1d22a0ca9_dictUmbrella.h) [C:\build\workspace\root-pullrequests-build\build\core\thread\G__Thread.vcxproj]
12:09:56    G__multimap2Dict.vcxproj -> C:\build\workspace\root-pullrequests-build\build\core\clingutils\G__multimap2Dict.dir\Release\G__multimap2Dict.lib
12:09:56    G__unordered_mapDict.vcxproj -> C:\build\workspace\root-pullrequests-build\build\core\clingutils\G__unordered_mapDict.dir\Release\G__unordered_mapDict.lib
12:09:57    G__forward_listDict.vcxproj -> C:\build\workspace\root-pullrequests-build\build\core\clingutils\G__forward_listDict.dir\Release\G__forward_listDict.lib
12:09:57  Command exited with the value: 1
12:09:57  MakeCommand:"C:\Program Files\CMake\bin\cmake.exe" --build . --config "Release" -- /maxcpucount:16
12:09:57  Error(s) when building project
12:09:57     2 Compiler errors
```

**Comment by bendavid** [comment]
Currently the usage of TBB in core/thread is triggered by the availability of tbb in the build (standalone or builtin).  Should the criteria be something different/more selective?  (or is there some other switch which needs to be triggered in the windows build for this case?)

**Comment by bendavid** [comment]
Alternatively this could be steered by whether the build is done with implicit multithreading enabled, rather than directly checking the availability of TBB (since this seems to not be sufficient in the windows case)  Thoughts?

**Comment by pcanal** [comment]
I suspect that instead we should still rely on detecting the availability of TBB *but* disable this feature on Windows, at least for now. (I.e. the condition is "has a TBB and not on Windows").

**Comment by bendavid** [comment]
Alright, done.

File: core/thread/test/CMakeLists.txt [hunk 1]
```diff
@@ -4,8 +4,16 @@
 # For the licensing terms see $ROOTSYS/LICENSE.
 # For the list of contributors see $ROOTSYS/README/CREDITS.
 
-ROOT_ADD_UNITTEST_DIR(Core Thread Hist)
+if (tbb OR builtin_tbb)
+   ROOT_ADD_UNITTEST_DIR(Core Thread Hist ${TBB_LIBRARIES})
 
-ROOT_ADD_GTEST(testTThreadedObject testTThreadedObject.cxx LIBRARIES Hist)
+   ROOT_ADD_GTEST(testTThreadedObject testTThreadedObject.cxx LIBRARIES Hist ${TBB_LIBRARIES})
```
Comment (pcanal) [inline core/thread/test/CMakeLists.txt:10]: Why is this needed?  Isn't tbb already linked through libThread (and/or libImt)?

**Comment by pcanal** [comment]
@phsft-bot build with flags -DCTEST_TEST_EXCLUDE_NONE=On

File: core/thread/test/CMakeLists.txt (hunk 1 above)
Comment (bendavid) [reply core/thread/test/CMakeLists.txt:10]: Yes I think you're right, will remove it.

**Comment by bendavid** [comment]
Updated to remove the unnecessary bits from the CMakeLists for the tests.

**Comment by pcanal** [comment]
@bendavid I think this is good to go except for adding some trace in the repository (either commit or, probably better, in the release notes) of the performance improvement.

**Comment by bendavid** [comment]
Thanks, I propose to add the relevant information to the release notes in a separate PR (mainly to avoid an otherwise unnecessary rebase before merging)

**Comment by pcanal** [comment]
Fair enough :)

**Comment by pcanal** [comment]
In the subsequent PR, consider also applying the formatting recommendation at: https://travis-ci.org/github/root-project/root/jobs/763450064

**Review by pcanal — APPROVED** [review]
